Add USB CDC and Serial tests

fkjagodzinski opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->
Add `tests-usb_device-serial` with a few test cases:
* `CDC RX multiple bytes`,
* `CDC RX multiple bytes concurrent`,
* `CDC RX single bytes`,
* `CDC RX single bytes concurrent`,
* `CDC USB reconnect`,
* `CDC loopback`,
* `Serial USB reconnect`,
* `Serial getc`,
* `Serial printf/scanf`,
* `Serial line coding change`,
* `Serial terminal reopen`.


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Feature
    [ ] Breaking change

CC @c1728p9 @jamesbeyond

fkjagodzinski committed:
Tests: USB: Add tests for CDC class.

fkjagodzinski committed:
Tests: USB: Add tests for Serial class.

fkjagodzinski committed:
Tests: USB: Fix USB reconnect tests for Win hosts

cmonr commented:
@c1728p9 @jamesbeyond Mind taking a look?

c1728p9 commented on TESTS/host_tests/usb_device_serial.py:
```diff
@@ -0,0 +1,251 @@
+    """Exception raised by retry_fun_call()."""
+
+
+def retry_fun_call(fun, num_retries=3, retry_delay=0.0):
```
Why are retries needed?

fkjagodzinski replied:
Operations like finding the serial port and opening it have to be retried due to variable delays on the host side. You can check how this works on your machine after setting `verbose` in line https://github.com/fkjagodzinski/mbed-os/blob/082c63e16a11b14a25b9b327843d2e8b1d7db408/TESTS/host_tests/usb_device_serial.py#L55

c1728p9 commented on TESTS/host_tests/usb_device_serial.py:
```diff
@@ -0,0 +1,251 @@
+        Search is based on the id received from the device itself.
+        Raises RuntimeError if the device is not found.
+        """
+        vid, pid = (int(i, base=16) for i in str(usb_id_str).split(','))
```
Can the device serial number be used instead? The serial number can be made unique per device. These tests will likely be running in an environment with many boards connected. VID/PID may not be sufficient to determine the device you are connected to.

fkjagodzinski replied:
I'll look into this. This should be rather straight forward to implement.

fkjagodzinski replied:
Changed as requested -- using USB SN to identify the correct device on the host system.

c1728p9 commented on TESTS/usb_device/serial/main.cpp:
```diff
@@ -0,0 +1,538 @@
+    Case("Serial USB reconnect", test_serial_usb_reconnect),
+    Case("Serial terminal reopen", test_serial_term_reopen),
+    Case("Serial getc", test_serial_getc),
+    Case("Serial printf/scanf", test_serial_printf_scanf),
```
It may also be good to test `line_coding_changed` if you are testing USBCDC in this PR.

fkjagodzinski replied:
New test case added -- `Serial line coding change`.

c1728p9 reviewed (dismissed):
@fkjagodzinski were these tests passing for you? When I run `tests-usb_device-serial` the test case `CDC loopback` consistently fails.

0xc0170 commented:
@fkjagodzinski Please respond to the latest review

jamesbeyond commented:
@0xc0170, Filip is on holiday, he will continue this PR when back to the office next week

cmonr commented:
Is @fkjagodzinski back?

fkjagodzinski commented:
@cmonr, yes I'm back and will update this PR soon.

fkjagodzinski committed:
Tests: USB: Use the serial number to find device

Use the USB device SN instead of VID & PID to find the device port name
on the host system.

fkjagodzinski committed:
Tests: USB: Add Serial test for line coding change

fkjagodzinski commented:
@c1728p9, thanks for your feedback. :+1: These 2 new commits update the code according to your comments.

I'm now moving to `CDC loopback` failures.

fkjagodzinski committed:
Tests: USB: Wait before closing host's port

This delay eliminates the possibility of the device detecting the port
being closed when still waiting for data.

fkjagodzinski committed:
Tests: USB: Wait before transmitting data to host

Despite having ECHO and ECHOCTL POSIX lflags disabled by default by
pyserial, a delay is needed for host to properly handle data received
from USB serial/CDC device.
With no delay host would echo 0x00-0x31 characters prefixed with '^'.

fkjagodzinski commented:
The `CDC loopback` test case is fixed now. @c1728p9 could you take a look?

c1728p9 commented on TESTS/usb_device/serial/main.cpp:
```diff
@@ -0,0 +1,665 @@
+    while (!usb_serial.connected()) {
+        wait_ms(1);
+    }
+    wait_ms(TX_DELAY_MS);
```
Why is there a delay here?

fkjagodzinski replied:
The first `usb_serial.printf()` call from the for loop has to be delayed. More details can be found in a comment for `TX_DELAY_MS` macro.

c1728p9 replied:
Is this a problem with the host machine or the mbed board? If this is an mbed problem then it needs  to be root caused.

fkjagodzinski replied:
This behaviour is Linux specific, when tests run on Win host, I do not see characters echoed with `^` prefix.

c1728p9 commented on TESTS/usb_device/serial/main.cpp:
```diff
@@ -0,0 +1,665 @@
+{
+    USBCDC usb_cdc(false, USB_CDC_VID, USB_CDC_PID);
+    char usb_cdc_sn[USB_SN_MAX_LEN] = { };
+    usb_desc2str(usb_cdc.string_iserial_desc(), usb_cdc_sn, USB_SN_MAX_LEN);
```
This will return the same serial number for all connected devices so the host won't be able to determine the right one to connect to. If testing is running concurrently on CI it may get these boards mixed up.

Instead, could the host test create and send a serial number to use in the test?

fkjagodzinski replied:
Ok, I'll look into this.

fkjagodzinski replied:
Done. The USB device now uses a SN (UUID) received from host on test suite setup.

c1728p9 commented on TESTS/usb_device/serial/main.cpp:
```diff
@@ -0,0 +1,665 @@
+    TEST_ASSERT_TRUE(usb_cdc.configured());
+    TEST_ASSERT_TRUE(usb_cdc.ready());
+
+    wait_ms(USB_DISCONNECT_DELAY_MS);
```
Why is USB_DISCONNECT_DELAY_MS needed? `usb_cdc.wait_ready()` should be enough to ensure that the host opened the serial port.

fkjagodzinski replied:
Correct, no issue on the device side. Additional delay `USB_DISCONNECT_DELAY_MS` helps Windows hosts handle this test case correctly. My Linux machine does not need it. I'll look into this and add comments for clarification.

c1728p9 commented on TESTS/usb_device/serial/main.cpp:
```diff
@@ -0,0 +1,665 @@
+    TEST_ASSERT_FALSE(usb_cdc.configured());
+    TEST_ASSERT_FALSE(usb_cdc.ready());
+
+    wait_ms(USB_RECONNECT_DELAY_MS);
```
Why does this need to be 200ms? Only a 1ms delay should be needed between disconnect and reconned as explained in the [basic test](https://github.com/ARMmbed/mbed-os/blob/9e06f07ebe475803b4f3ddd6cd574e5e72293794/TESTS/usb_device/basic/main.cpp#L39).

fkjagodzinski replied:
Good catch! 1 ms delay works as expected.

c1728p9 commented on TESTS/usb_device/serial/main.cpp:
```diff
@@ -0,0 +1,665 @@
+{
+    USBCDC usb_cdc(false, USB_CDC_VID, USB_CDC_PID);
+    char usb_cdc_sn[USB_SN_MAX_LEN] = { };
+    usb_desc2str(usb_cdc.string_iserial_desc(), usb_cdc_sn, USB_SN_MAX_LEN);
```
You might want to send lots of data to the device as part of this test. Event if its more data then the device can hold at once, the built in flow control should ensure that nothing is dropped.

fkjagodzinski replied:
Done.

fkjagodzinski committed:
Tests: USB: Correct reconnect delay

Setting to 1 ms to be on the safe side. The USB spec defines this delay
to be at least 200 us.

fkjagodzinski committed:
Tests: USB: Use SN generated by the host machine

fkjagodzinski committed:
Tests: USB: Update code comments

fkjagodzinski committed:
Tests: USB: Fix Windows USB SN handling

fkjagodzinski committed:
Tests: USB: Increas the size of RX CDC test data

This ensures the flow control is tested.

cmonr commented:
@c1728p9 @jamesbeyond All good?

cmonr commented:
@c1728p9 @jamesbeyond Ping

fkjagodzinski commented:
@c1728p9 thanks for your feedback.

Although the `TX_DELAY_MS` value should matter only for Linux hosts, when I removed this delay, I found another issue present on Windows hosts. I suspect this to be a host side issue and think this PR should proceed. I'll update this test suite in another PR if it is needed.

c1728p9 committed:
Tests: USB: Use DTR to make Serial tests reliable

Update the serial host test to set DTR low before opening a serial
port. Set DTR high once the serial object is ready. This ensures that
no data is sent by the device until the host is ready.

c1728p9 committed:
Tests: USB: Remove TX_DELAY_MS from serial test

Remove TX_DELAY_MS and all the places it was called since it is not
needed when DTR flow control is used.

c1728p9 committed:
Tests: USB: Increase serial open retries

Double the number of attempts that are made to open a serial port.
This makes tests run more reliably on Windows.

c1728p9 commented:
Hi @fkjagodzinski I looked into the `TX_DELAY_MS` failures on Windows and found them to be a race-condition in pyserial. As part of pyserial's serial port `open` call it first opens the physical serial port and then does a [flush of TX and RX buffers](https://github.com/pyserial/pyserial/blob/master/serial/serialwin32.py#L84). Any data that is received between the open and flush is lost, causing tests like the loopback test to hang and timeout.

As a fix for this DTR (data terminal ready) can be set to False before opening the serial port. This prevents the device from sending any data. Once the port is opened and ready DTR can be re-enabled to allow testing like normal.

With this change I don't need any of the calls to `wait_ms(TX_DELAY_MS);` and I don't see any failures even across multiple runs. I pushed my changes https://github.com/c1728p9/mbed-os/tree/increase_stability. Feel free to make use of this in this or a subsequent PR.

c1728p9 committed:
Remove USB disconnect delay from Serial test

Remove the USB disconnect delay since there is no reason this is
needed. Failures which occur without this indicate a device or host
problem.

fkjagodzinski committed:
Tests: USB: DTR fix for Linux hosts

A DTR line is used to signal that the host has configured a terminal and
is ready to transmit and receive data from the USB CDC/Serial device.
When this test suite is run with the use of a Linux host, a workaround has
to be used to overcome some platform specific DTR line behavior.
Every time the serial port file descriptor is opened, the DTR line is
asserted until the terminal attributes are set.
As a consequence, the device receives a premature DTR signal with a
duration of 200-500 us before the correct, long-lasting DTR signal set by
the host-side test script. (tested on the Linux kernel 4.15.0)

The solution is to wait for the first DTR spike, ignore it, and wait for
the correct DTR signal again.

fkjagodzinski commented:
* cherry-picked a few fixes from @c1728p9,
* added a fix to handle the Linux hosts correctly (regarding the DTR line)

c1728p9 commented on TESTS/host_tests/usb_device_serial.py:
```diff
@@ -0,0 +1,346 @@
+
+    def port_open_wait(self):
+        """Open the serial and wait until it's closed by the device."""
+        mbed_serial = serial.Serial(dsrdtr=False)
```
Why is ```dsrdtr``` set to false here and below?

fkjagodzinski replied:
This is just setting the default `dsrdtr` value explicitly. It has to stay `False` to allow setting the `DTR` line manually, i.e. setting `DTR` `True` after the call to `open()`.

fkjagodzinski commented:
Branch rebased to current `feature-hal-spec-usb-device`.

fkjagodzinski commented on TESTS/usb_device/serial/main.cpp:
```diff
@@ -0,0 +1,847 @@
+//
+// The solution is to wait for the first DTR spike, ignore it, and wait for
+// the correct DTR signal again.
+#define LINUX_HOST_DTR_FIX 1
```
Following a @donatieng suggestion, I checked how the USB gadget ACM driver might handle this case, and well.. it does not respect DTR state at all. :confused: 

https://github.com/torvalds/linux/blob/84df9525b0c27f3ebc2ebb1864fa62a97fdedb7d/drivers/usb/gadget/function/f_acm.c#L387-L392

0xc0170 commented:
@fkjagodzinski What is the status? It has not yet been approved. Is there anything missing?

fkjagodzinski commented:
Waiting for a final approval from @c1728p9.

c1728p9 commented:
@fkjagodzinski have you tried this on OSX? If not you may want to ensure there are no problems there.

c1728p9 reviewed (approved):
This PR looks good to me. Thanks @fkjagodzinski

fkjagodzinski commented:
> @fkjagodzinski have you tried this on OSX? If not you may want to ensure there are no problems there.

Good point. Unfortunately I don't have access to any OSX machine now.

cmonr commented:
/morph build

fkjagodzinski commented:
The USB device tests require an additional USB cable connected to the host. These tests should not run on the CI until additional cables are provided.

@cmonr, @0xc0170, @studavekar, could you update the CI config

https://github.com/ARMmbed/austin-mbed-ci/blob/9d0b962742d574cacb1b7c2cc1a4364d8819a3c0/jobs/test_matrix_raas/config.xml#L344-L345

and replace `tests-usb_device-basic` with `tests-usb_device-*`, or should I create a PR to that repo?

0xc0170 commented:
Please send PR, and notify @studavekar

studavekar commented:
@c1728p9, @fkjagodzinski HW test CI config is updated to filter out new USB tests.

fkjagodzinski commented:
Great! Thanks @studavekar.

cmonr commented:
/morph test

cmonr commented:
Will restart in a bit. Figured we had a spurious network issue, when this is only waiting on a test.

cmonr commented:
/morph test

cmonr commented:
/morph test

0xc0170 commented:
/morph test

0xc0170 commented:
/morph test

adbridge commented:
/morph test

c1728p9 commented:
@fkjagodzinski the test case `Serial printf/scanf` fails when IAR is used. Can you investigate and correct this failure?

fkjagodzinski commented:
> @fkjagodzinski the test case `Serial printf/scanf` fails when IAR is used. Can you investigate and correct this failure?

This should be fixed with #9033.

TomoYamanaka commented:
@fkjagodzinski  @c1728p9 
Thank you so much. I confirmed that the above test case is success with IAR by adopting #9033 .

I'm preparing to submit a PR for Renesas' s USB Device feature, so I'll be  glad if you could rebase `feature-hal-spec-usb-device` repository to merge #9033 .
